security-csrf/src/main/java/io/micronaut/security/csrf/generator/DefaultCsrfTokenGenerator.java

```diff
@@ -0,0 +1,98 @@
+import io.micronaut.http.cookie.CookieConfiguration;
+import io.micronaut.security.csrf.CsrfConfiguration;
+import io.micronaut.security.session.SessionIdResolver;
+import io.micronaut.security.utils.HMacUtils;
+import jakarta.inject.Singleton;
+
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.Base64;
+
+/**
+ * Default implementation of {@link CsrfTokenGenerator} which generates a CSRF Token prefixed by an HMAC if a secret key is set.
+ * @see <a href="https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#pseudo-code-for-implementing-hmac-csrf-tokens">Pseudo Code for implementing hmac CSRF tokens</a>
+ * @author Sergio del Amo
+ * @since 4.11.0
+ * @param <T> Request
+ */
+@Requires(classes = CookieConfiguration.class)
+@Singleton
+public final class DefaultCsrfTokenGenerator<T> implements CsrfTokenGenerator<T> {
+    /**
+     * hmac random value separator.
+     */
+    public static final String HMAC_RANDOM_SEPARATOR = ".";
+    private static final String SESSION_RANDOM_SEPARATOR = "!";
+    private final SecureRandom secureRandom = new SecureRandom();
+    private final CsrfConfiguration csrfConfiguration;
+    private final SessionIdResolver<T> sessionIdResolver;
+
+    /**
+     *
+     * @param csrfConfiguration CSRF Configuration
+     * @param sessionIdResolver SessionID Resolver
+     */
+    public DefaultCsrfTokenGenerator(CsrfConfiguration csrfConfiguration,
+                              SessionIdResolver<T> sessionIdResolver) {
+        this.csrfConfiguration = csrfConfiguration;
+        this.sessionIdResolver = sessionIdResolver;
+    }
+
+    @Override
+    @NonNull
+    public String generateCsrfToken(@NonNull T request) {
+        byte[] tokenBytes = new byte[csrfConfiguration.getRandomValueSize()];
+        secureRandom.nextBytes(tokenBytes);
+        String randomValue = Base64.getUrlEncoder().withoutPadding().encodeToString(tokenBytes);  // Cryptographic random value
+
+        String hmac = hmac(request, randomValue);
+        // Add the `randomValue` to the HMAC hash to create the final CSRF token. Avoid using the `message` because it contains the sessionID in plain text, which the server already stores separately.
+        return  hmac + HMAC_RANDOM_SEPARATOR + randomValue;
+    }
+
+    /**
+     *
+     * @param request Request
+     * @param randomValue Cryptographic random value
+     * @return HMAC hash
+     */
+    @NonNull
+    public String hmac(@NonNull T request, String randomValue) {
+        // Gather the values
+        String secret = csrfConfiguration.getSecretKey();
+        String sessionID = sessionIdResolver.findSessionId(request).orElse(""); // Current authenticated user session
+
+        // Create the CSRF Token
+        String message = sessionID + SESSION_RANDOM_SEPARATOR + randomValue; // HMAC message payload
```

Comment:
Is it necessary to use this format? Otherwise, to prevent attacks against the format, I suggest

```suggestion
        String message = sessionID.length() + "." + sessionID + "." + randomValue.length() + "." + randomValue; // HMAC message payload
```

Reply:
The problem is described in more detail in the first section here: https://blog.trailofbits.com/2024/08/21/yolo-is-not-a-valid-hash-construction/

Reply:
the CSRF token generation follows the [OWSAP example](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#pseudo-code-for-implementing-hmac-csrf-tokens)

Reply:
resolved via https://github.com/micronaut-projects/micronaut-security/pull/1826/commits/d8d5020af2080ccb7a45bcc857751164dcde65c0